lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +40,57 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(config: PbenchServerConfig, logger: Logger):
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server need to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+    """
+    try:
+        oidc_server = str(config.get("authentication", "internal_server_url"))
+    except (NoSectionError):
+        logger.exception("Error fetching required configuration")
+        sys.exit(1)
+    except (NoOptionError):
+        try:
+            oidc_server = str(config.get("authentication", "server_url"))
+        except (NoOptionError):
+            logger.exception("Error fetching required configuration")
+            sys.exit(1)
+
+    # The connection check will run for a minute, and if the connection is
+    # successful it will break the look.
+    t_end = time.time() + 60
+    while time.time() <= t_end:
+        try:
+            response = requests.session().request(
+                "GET", f"{oidc_server}/health", verify=False
+            )
+            response.raise_for_status()
+            if response.json()["status"] == "UP":
+                return
+            else:
+                time.sleep(10)
+                continue
+        except Exception:
+            logger.exception("Exception while initializing OIDC client")
```

Comment:
I'm not sure that we want to use `logger.exception()` here:  we almost certainly don't care about the traceback; and, given that we're going to retry the request, we pretty much don't care about the exception (that is, receiving an exception here is not necessarily "exceptional").

I might be inclined to capture the reason why we failed on each trip through the loop (whether it's due to an exception or to a status other than `UP`), and then report that if we time out, but I'm not sure I would bother with reporting failures every 10 seconds, especially if we manage to start up successfully.

(To be honest, I probably wouldn't use `raise_for_status()` -- it's not that hard to check for `200` or `response.OK` and then check for the status -- and then skip using the `try` block altogether.)

Reply:
Using the urllib3 Retry and python request HTTPAdapter instead of while loop and try-except block.